Show Carbs required - Updated from #1974

Tornado-Tim opened the pull request:
This PR Updates #1974 and the awesome work that @twain47 did, merges it with the recent oref update and then brings it up to date with dev.

Some refactoring has been done with migrating some components to Kotlin and using dagger.

![Screenshot (3 May 2020 16_06_10)](https://user-images.githubusercontent.com/2896311/80907191-d8b3d200-8d68-11ea-9a1e-b8d48140a443.png)
![Screenshot (3 May 2020 14_39_12)](https://user-images.githubusercontent.com/2896311/80907192-dbaec280-8d68-11ea-8836-e76cc7fca2a5.png)

MilosKozak commented on app/src/main/res/values/strings.xml:
```diff
@@ -194,6 +194,8 @@
     <string name="lgs">LGS</string>
     <string name="openloop_newsuggestion">New suggestion available</string>
+    <string name="carbssuggestion">Carbs Suggestion</string>
+    <string name="unsupportedclientver">Unsupported version of NSClient</string>
```
Added back by mistake or merge

MilosKozak commented on app/src/main/res/values/strings.xml:
```diff
@@ -774,10 +777,14 @@
     <string name="smb_shortname">SMB</string>
     <string name="key_use_smb" translatable="false">use_smb</string>
     <string name="key_use_uam" translatable="false">use_uam</string>
+    <string name="key_smb_enable_carbs_suggestions" translatable="false">key_smb_enable_carbs_suggestions</string>
```
2x
content without key_

MilosKozak commented:
Generaly great.
What is a bit against current logic is not utilizing internal Notification. 
Someone prefers it on main screen someone prefers conversion to Android notification.

So I'd recommend improve internal Notification for such usage and let it convert to adnroid if prefered by user

swissalpine commented:
Thanks again for your work!
I have already benefited from the update to oref0.7 for quite some time, now from this update. I can confirm: It works!

![image](https://user-images.githubusercontent.com/12517693/81483357-e7552a00-923d-11ea-98e0-b1759c7ff494.png)

Tornado-Tim commented:
Updated the PR @MilosKozak to allow local alerting, and Uploading to NS the announcement of carbs required.
Other changes now include the carb Icon pulsing red when carbs are required with the required amount underneath.
![Screenshot (10 May 2020 20_36_08)](https://user-images.githubusercontent.com/2896311/81495582-66f9fc00-9305-11ea-9388-994607452ce7.png)

MilosKozak commented:
👍

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -184,6 +187,11 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
         overview_bggraph?.gridLabelRenderer?.reloadStyles()
         overview_bggraph?.gridLabelRenderer?.labelVerticalWidth = axisWidth
 
+        val carbicon = overview_carbs_icon;
```
this seems to just copy a reference.
I believe you could simply use it directly in the next lines.

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -184,6 +187,11 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
 
+        val carbicon = overview_carbs_icon;
+        carbAnimation = carbicon.background as AnimationDrawable
+        carbAnimation.setEnterFadeDuration(1200);
```
There are still `;` from transformed Java code.

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -733,9 +741,19 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
+            overview_cob?.text = cobText + " ("+lastRun?.constraintsProcessed.carbsReq + " req)"
+            if(!carbAnimation.isRunning)
+                carbAnimation.start()
+        }else{
```
Could you please hit command+shift+L (reformat). (And hopefully this doesn't show where we have other formatting issues ;))

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -733,9 +741,19 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
 
         val lastRun = loopPlugin.lastRun
+        if (config.APS && lastRun?.constraintsProcessed != null && lastRun?.constraintsProcessed.carbsReq > 0) {
+            overview_cob?.text = cobText + " ("+lastRun?.constraintsProcessed.carbsReq + " req)"
```
shouldn't we have another marker than "()" that is used for future carbs already -  or better "req" internationalized?

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -733,9 +741,19 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
+                carbAnimation.start()
+        }else{
+            overview_cob?.text = cobText
+            if(carbAnimation.isRunning)
```
I'm not entirely sure about the animation part. It breaks a bit with the other UI.
But that's not my topic, so I go with what @MilosKozak and @TebbeUbben deem good.

Tornado-Tim replied:
OK, I think the animation brings awareness to what needs attention.

swissalpine replied:
Hi Tim,
I disabled in "my" version the animation (a question of my personally taste) and colored the cob text in yellow. I realised the following writing:
32 g (52 g) | 12 g req
But that is only my negligible thoughts.

Tornado-Tim replied:
Yea issue is here we don't have enough space on the main screen anymore to show the text without it getting too in your face.
Not entirely happy about the Cob required being underneath the Carb icon as it makes it too small.

I brought the animation in as CarbsReq is a situation that needs highlighting, its hard to highlight importance but not get your face at the same time.

AdrianLxM commented on app/src/main/java/info/nightscout/androidaps/plugins/aps/loop/LoopPlugin.java:
```diff
@@ -431,6 +436,67 @@ public synchronized void invoke(String initiator, boolean allowNotification, boo
+                            intentAction15m.putExtra("ignoreDuration", 15);
+                            PendingIntent pendingIntent15m = PendingIntent.getBroadcast(context, 1, intentAction15m, PendingIntent.FLAG_UPDATE_CURRENT);
+                            NotificationCompat.Action actionIgnore15m = new
+                                    NotificationCompat.Action(R.drawable.ic_notif_aaps, "Ignore 15m", pendingIntent15m);
```
Would be great if "Ignore (5|15|30)m" could be localized.

swissalpine replied:
I see two messages on my watch.
1. A notification with carbs required with the possibility to dismiss.
2. A second notification with the ignore possibilities ...

Tornado-Tim replied:
> I see two messages on my watch.
> 
> 1. A notification with carbs required with the possibility to dismiss.
> 2. A second notification with the ignore possibilities ...

Point 2: Can you elaborate this for me "ignore possibilities", as that text reference isnt in this PR. Did you enable the option to upload to NS, Carb required announcements? as that will have double notification. 
(one for the carbs being required, then another from NSclient uploading then processing the carbs required announcement)

swissalpine replied:
Sorry for my englisch: I meant with "ignore possibilities": "Ignore 15m" and so on.
This is the notification I used to see with the first PR.
Now I see another message just before. And no, it should not come from Nightscout. I haven't activated the option and couldn't see a carb required notice in Nightscout.

swissalpine replied:
Remark of a dummy: Maybe the messages are generated by the double function in LoopPlugin.java
a) local notification
b) android notification
To avoid the double probably call the local notification could not be sent if android notifications are allowed ...

swissalpine replied:
I was thinking something like this:
```
445 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && !sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)) {

452 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)){
```
Please lough if this is stupid ...

swissalpine replied:
Can confirm that this lines fixes the double notification problem.

MilosKozak commented on app/src/main/java/info/nightscout/androidaps/plugins/general/overview/OverviewFragment.kt:
```diff
@@ -748,14 +747,14 @@ class OverviewFragment : DaggerFragment(), View.OnClickListener, OnLongClickList
         if (config.APS && lastRun?.constraintsProcessed != null && lastRun?.constraintsProcessed.carbsReq > 0) {
-            overview_cob?.text = cobText + " ("+lastRun?.constraintsProcessed.carbsReq + " req)"
-            if(!carbAnimation.isRunning)
+            overview_cob?.text = cobText + " *" + lastRun?.constraintsProcessed.carbsReq + " "+R.string.required+"*"
```
resourceHelper.gs(R.string.required)

MilosKozak commented on app/src/main/java/info/nightscout/androidaps/plugins/aps/loop/LoopPlugin.java:
```diff
@@ -431,6 +436,67 @@ public synchronized void invoke(String initiator, boolean allowNotification, boo
+                            intentAction5m.putExtra("ignoreDuration", 5);
+                            PendingIntent pendingIntent5m = PendingIntent.getBroadcast(context, 1, intentAction5m, PendingIntent.FLAG_UPDATE_CURRENT);
+                            NotificationCompat.Action actionIgnore5m = new
+                                    NotificationCompat.Action(R.drawable.ic_notif_aaps, sp.getString(R.string.ignore5m,"Ignore 5m"), pendingIntent5m);
```
sp.getString -> resourceHelper.gs

swissalpine commented:
I don't know if you read my proposal above. The double notification fixes the change of the following program lines in LoopPlugin.java:
```
445 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && !sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)) {

452 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)){
```

Another small problem is that immediately after a carbohydrate input and insulin delivery via the calculator a notification is sent, usually about the amount just entered. Is it possible to write a snooze function that switches off the notification for e.g. 15 minutes after carbohydrate/insulin delivery?

Thanks for your work!

MilosKozak commented on app/src/main/jniLibs/README.md:
```diff
@@ -0,0 +1 @@
+# danars-support-lib
```
this + `so` objects are added by mistake

Tornado-Tim commented:
> I don't know if you read my proposal above. The double notification problem is fixed for me with the change of the following program lines in LoopPlugin.java:
> 
> ```
> 445 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && !sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)) {
> 
> 452 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)){
> ```
> 
> Another small problem is that immediately after a carbohydrate input and insulin delivery via the calculator a notification is sent, usually about the amount just entered. Is it possible to write a snooze function that switches off the notification for e.g. 15 minutes after carbohydrate/insulin delivery?
> 
> Thanks for your work!

Hey @swissalpine, thanks for your input. In the above situation it would make notifications exclusive local or Android, which is something I am not sure we would want as people would find that some of there notification settings are not working.

You shouldnt be betting double notifications if you turn the local one off in the settings.

This conversation about notifications (Android vs local) is something I have raised with the team.
However this whole conversation is bit more broad than just Carb required notifications.

If you could indulge me and reverting your patch in the meantime and test turning off your local carb notifications. When that setting is off and Android notifications are on, you should not see 2 notifications at the same time.

I will have a look at the snooze function in the meantime.

Tornado-Tim commented:
> > I don't know if you read my proposal above. The double notification problem is fixed for me with the change of the following program lines in LoopPlugin.java:
> > ```
> > 445 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && !sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)) {
> > 
> > 452 if (sp.getBoolean(R.string.key_enable_carbs_required_alert_local,true) && sp.getBoolean(R.string.key_raise_notifications_as_android_notifications, false)){
> > ```
> > 
> > 
> > Another small problem is that immediately after a carbohydrate input and insulin delivery via the calculator a notification is sent, usually about the amount just entered. Is it possible to write a snooze function that switches off the notification for e.g. 15 minutes after carbohydrate/insulin delivery?
> > Thanks for your work!
> 
> Hey @swissalpine, thanks for your input. In the above situation it would make notifications exclusive local or Android, which is something I am not sure we would want as people would find that some of the notification settings are not working.
> 
> You shouldnt be be getting double notifications if you turn the local one off in the settings.
> 
> This conversation about notifications (Android vs local) is something I have raised with the team.
> However this whole conversation is bit more broad than just Carb required notifications.
> 
> If you could indulge me and reverting your patch in the meantime and test turning off your local carb notifications. When that setting is off and Android notifications are on, you should not see 2 notifications at the same time.
> 
> I will have a look at the snooze function in the meantime.

OK, I have got clarification on this now. Notifications should be either Android or Local.

Will update the PR to reflect this.

swissalpine commented:
Thanks!
I've always assumed that: First I define which alarms should be active. These are then initially local alarms. If I have them as system notifications, I set an additional check mark with the consequence that the alarms (previously marked as local) become system notifications.
--
What do you think about the snooze function. Some minutes after entering carbs und insulin I get a carbs suggestion ... It's a little annoying - but overall you've made a great expansion. Thanks again for that!

Tornado-Tim commented:
Added a snooze function of 15 min post treatment entry (carb or bolus entry).  there is scope to make this a setting, but AAPS has so many settings, I have set the default to 15 which is a good middle ground.

Added CarbTime functions and did some minor refractoring so other fuctions would use the Carbtime function.

swissalpine commented:
Thanks! I will give it a try.

swissalpine commented:
The snooze function works perfect. I don't think there's a need for a special setting.

ToreBj commented:
Thank you, This seems to work very well. I also get notifications. I like the glowing symbol going to red and back :) Good work!
![Screenshot_20200615-200935](https://user-images.githubusercontent.com/23278852/84878887-c4166980-b08a-11ea-98c2-bd3772df84fc.jpg)
